### PR TITLE
[Breaking] Reintroduce Bundle, Finalize Names

### PR DESCRIPTION
**Pull request by Kashomon**
This PR is a follow-up on the PR to convert to Go structs from Proto and
contains several schema modifications, additional tests, and usability
helpers.

- Reintroduce the Bundle type. This PR reintroduces the Bundle as a K8S type,
  but now with much greater clarity about what it is. Now, the Bundle is
  essentially a renamed `ComponentData` (existed previously) but with the 
  additional type and   object meta information that being a K8S type provides. 
  Additionally, this type includes comments that indicate this is meant for release
  tooling and libraries, and so, no client generation is included with this
  type.
- Add a name to ComponentSet and Bundle, and call it `SetName`. We've
  had several discussions about what the right name for this is. I've
  called it simply `SetName` in both cases. I would have loved to call it
  GroupName, but that moniker is already taken.
- Rename the name field in ComponentPackage to `ComponentName`. This is much
  clearer than `CanonicalName`, albeit a bit long. but it is simple, and I think clear, 
  which I like.
- Rework ComponentSet to use a 'ComponentReference' object that contains the
  ComponentName and the Component Version.
- Add helpers to convert between all these types:
  - Add a helper to create LocalObjectReference objects from the
    ComponentReference object
  - Add a helper to create standardized metadata.name fields for all objects that have 
  either ComponentName or SetName + Version
  - Add a helper to generate a component set from a Bundle.
- Now that names are settling down, add a bunch of tests to ensure that the
  name fields match the K8S standards on naming.

### CODE REVIEW

**Comment by jplobosgle** [comment]
I couldn't do a close review, given its size, but the ideas make sense. 

/lgtm

File: pkg/validation/validate_name.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+var firstCharRegexp = regexp.MustCompile(`^[a-z0-9]`)
+var nameRegexp = regexp.MustCompile(`^[a-z0-9_.-]+$`)
+var lastCharRegexp = regexp.MustCompile(`[a-z0-9]$`)
+
+// ValidateName validates names to ensure they follow the Kubernetes
+// conventions for how names are constructed. For more about names,
+// see: k8s.io/docs/concepts/overview/working-with-objects/names/
+func ValidateName(n string) error {
```
Comment (mtaufen) [inline pkg/validation/validate_name.go]: I didn't compare them but you may want to look at `k8s.io/kubernetes/pkg/apis/core/validation/validation.go:ValidateObjectMeta` and see if this function covers the same properties as name validation in there.

**Comment by ihmccreery** [comment]
Please prioritize #89.  Once that's done, this lgtm.

File: pkg/apis/bundle/v1alpha1/bundle_types.go [hunk 2]
```diff
@@ -30,101 +42,12 @@ type ComponentSetSpec struct {
...
-	// the file is YAML-formatted and contains multiple objects in the YAML-doc,
-	// the objects will be split into separate inline objects. In other words,
-	// one object file may result in multiple objects.
-	//
-	// Each object file must be parsable into a Struct: In other words,
-	// it should be representable as either YAML or JSON.
-	ObjectFiles []File `json:"objectFiles,omitempty"`
+	// ComponentFiles reference ComponentPackage files. The component files can
+	// be inlined as directly included components (see `Components` below).
+	ComponentFiles []*File `json:"componentFiles,omitempty"`
```
[the remark this answers is not in the document]
Comment (mtaufen) [reply pkg/apis/bundle/v1alpha1/bundle_types.go]:
```go
// A
c := &Components[i]
c.X = val

// B
Components[i].X = val
``` 
Are both explicit ways to change a substructure in-place, without adding pointer indirection to the type.

**Pull request merged**